### PR TITLE
Update and restructure rules documentation

### PR DESCRIPTION
**Pull request by ralphje**
The documentation on Suricata Rules could use some improvement, mainly because the ToC levels varied across sections. I tried to improve the documentation as follows:

* Removed screenshots of rules in first few sections (ongoing effort)
* Rewritten the introduction and parts of meta keywords
* Lowercased all section titles that are keywords
* Moved explanation of modifiers and normalization to the introduction page
* Moved fast_pattern and prefilter to their own page as they do similar things
* Moved pcre from its dedicated page to the payload keywords page, and removed the reference to it from the HTTP keywords page
* Moved flowint from its dedicated page to the flow keywords page.
* Moved IP reputation keywords to the rules section.
* Restructured documentation to have more consistent ToC levels

This is a resubmission of PR #2927 with cleaner and more separated commits. I didn't split this out to two PRs since this made it a lot easier for me to ensure all changes were submitted, and both PRs would build on each other.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

### CODE REVIEW

File: doc/userguide/reputation/ipreputation/ip-reputation-format.rst [hunk 1]
```diff
@@ -1,7 +1,7 @@
 IP Reputation Format
 ====================
 
-Description of IP Reputation file formats. For the configuration see :doc:`ip-reputation-config` and :doc:`ip-reputation-rules` for the rule format.
+Description of IP Reputation file formats. For the configuration see :doc:`ip-reputation-config` and :doc:`/rules/ip-reputation-rules`  for the rule format.
```
Comment (norg) [inline doc/userguide/reputation/ipreputation/ip-reputation-format.rst]: unnecessary space? :)

File: doc/userguide/rules/intro.rst [hunk 2]
```diff
@@ -30,172 +32,197 @@ In this example, :example-rule-action:`red` is the action,
...
+There are also a few so-called application layer protocols, or layer 7 protocols
+you can pick from. These are:
 
-You can not write a signature using EXTERNAL_NET because it stands for
-'not any'. This is a invalid setting.
+* http
+* ftp
+* tls (this includes ssl)
+* smb
+* dns (from v2.0)
```
Comment (norg) [inline doc/userguide/rules/intro.rst]: we also have smtp for example, icmp and ssh are also in ET rulesets.

File: doc/userguide/rules/intro.rst [hunk 3]
```diff
@@ -30,172 +32,197 @@ In this example, :example-rule-action:`red` is the action,
...
 .. container:: example-rule
 
     drop tcp $HOME_NET :example-rule-emphasis:`any` -> $EXTERNAL_NET :example-rule-emphasis:`any` (msg:"ET TROJAN Likely Bot Nick in IRC (USA +..)"; flow:established,to_server; flowbits:isset,is_proto_irc; content:"NICK "; pcre:"/NICK .*USA.*[0-9]{3,}/i"; reference:url,doc.emergingthreats.net/2008124; classtype:trojan-activity; sid:2008124; rev:2;)
 
 *The first emphasized part is the source, the second is the destination (note the direction of the directional arrow).*
 
-In this example, the red, bold-faced part is the port.
+Traffic comes in and goes out through ports. Different ports have
+different port numbers. For example, the default port for HTTP is 80 while 443 is
+the typically the port for HTTPS. Note, however, that the port does not
```
Comment (norg) [inline doc/userguide/rules/intro.rst]: double "the" so I sugges "443 is typically the port for HTTPS"

File: doc/userguide/rules/intro.rst (hunk 2 above)
Comment (ralphje) [reply doc/userguide/rules/intro.rst]: Are these keywords supported for the protocol field as well? As there's no list anywhere.

**Comment by victorjulien** [comment]
The travis-ci warning can be ignored. It should go away after #2937 is merged.

File: doc/userguide/rules/intro.rst (hunk 2 above)
Comment (norg) [reply doc/userguide/rules/intro.rst]: Yes, those are at the protocol field. We have some more, see also https://suricata-ids.org/features/all-features/ but adding that in a seperate commit/PR is totally fine with me.

**Comment by victorjulien** [comment]
<pre>
Warning, treated as error:
/home/victor/sync/devel/eidps/doc/userguide/rules/header-keywords.rst:415: ERROR: Malformed table.
Column span alignment problem in table line 5.

==========  ==========  ==========================================================
ICMP Code   ICMP Type   Description
==========  ==========  ==========================================================
3           - 0         - Net Unreachable
            - 1         - Host Unreachable
            - 2         - Protocol Unreachable
            - 3         - Port Unreachable
            - 4         - Fragmentation Needed and Don't Fragment was Set
            - 5         - Source Route Failed
            - 6         - Destination Network Unknown
            - 7         - Destination Host Unknown
            - 8         - Source Host Isolated
            - 9         - Communication with Destination Network is Administratively Prohibited
            - 10        - Communication with Destination Host is Administratively Prohibited
            - 11        - Destination Network Unreachable for Type of Service
            - 12        - Destination Host Unreachable for Type of Service
            - 13        - Communication Administratively Prohibited
            - 14        - Host Precedence Violation
            - 15        - Precedence cutoff in effect
5           - 0         - Redirect Datagram for the Network (or subnet)
            - 1         - Redirect Datagram for the Host
            - 2         - Redirect Datagram for the Type of Service and Network
            - 3         - Redirect Datagram for the Type of Service and Host
9           - 0         - Normal router advertisement
            - 16        - Doest not route common traffic
11          - 0         - Time to Live exceeeded in Transit
            - 1         - Fragment Reassembly Time Exceeded
12          - 0         - Pointer indicates the error
            - 1         - Missing a Required Option
            - 2         - Bad Length
40          - 0         - Bad SPI
            - 1         - Authentication Failed
            - 2         - Decompression Failed
            - 3         - Decryption Failed
            - 4         - Need Authentication
            - 5         - Need Authorization
==========  ==========  ==========================================================

Makefile:616: recipe for target '_build/latex/Suricata.pdf' failed
make[2]: *** [_build/latex/Suricata.pdf] Error 1
make[2]: Leaving directory '/home/victor/sync/devel/eidps/doc/userguide'
Makefile:519: recipe for target 'pdf-recursive' failed
make[1]: *** [pdf-recursive] Error 1
make[1]: Leaving directory '/home/victor/sync/devel/eidps/doc'
Makefile:494: recipe for target 'pdf-recursive' failed
make: *** [pdf-recursive] Error 1
</pre>
Getting a warning, could you check what is happening?

**Comment by ralphje** [comment]
Doesn't happen on my end, but should be fixed now.

**Comment by victorjulien** [comment]
Now I get:
Warning, treated as error:
/home/victor/sync/devel/eidps/doc/userguide/rules/fast-pattern.rst:: WARNING: document isn't included in any toctree

**Comment by ralphje** [comment]
Why did you close it? 
I haven't yet had the time to look into your issue due to holiday

**Comment by victorjulien** [comment]
On 20-11-17 13:02, Ralph Broenink wrote:
> Why did you close it?
> I haven't yet had the time to look into your issue due to holiday

It needs work and our workflow explains that after feedback we need a
new branch+PR. See
https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Github_work_flow#Create-a-new-branch-for-incorporating-feedback

The PR list is basically a todo-list for me, so I will remove things
that need to be worked on by others.

Thanks!

**Comment by ralphje** [comment]
I don't get that same error. How do you build the documentation?

**Comment by victorjulien** [comment]
On 20-11-17 13:46, Ralph Broenink wrote:
> I don't get that same error. How do you build the documentation?

In doc/userguide I do 'make pdf'.

**Comment by ralphje** [comment]
The pdf target does not exist: ```make: *** No rule to make target `pdf'.  Stop.```

If you mean `make latexpdf`, I don't have LaTeX installed, but running `make latex` succeeds without any issue (and that's the part where you probably get the error.)

Does `make html` work without error?

**Comment by victorjulien** [comment]
Yes the pdf target does exist:
<pre>
$ make pdf
sysconfdir=/etc \
localstatedir=/var \
version=4.0.0-dev \
	sphinx-build -q -W -b latex -d _build/doctrees \
	../../doc/userguide _build/latex

Warning, treated as error:
/home/victor/sync/devel/eidps/doc/userguide/rules/fast-pattern.rst:: WARNING: document isn't included in any toctree

Makefile:616: recipe for target '_build/latex/Suricata.pdf' failed
make: *** [_build/latex/Suricata.pdf] Error 1
</pre>
html:
<pre>
$ make html
sysconfdir=/etc \
localstatedir=/var \
version=4.0.0-dev \
	sphinx-build -q -W -b html -d _build/doctrees \
	../../doc/userguide _build/html

Warning, treated as error:
/home/victor/sync/devel/eidps/doc/userguide/rules/fast-pattern.rst:: WARNING: document isn't included in any toctree

Makefile:609: recipe for target 'html' failed
make: *** [html] Error 1
</pre>

**Comment by ralphje** [comment]
Which version of sphinx-build do you have?

**Comment by victorjulien** [comment]
Sphinx v1.3.6 (default of Ubuntu 1604)

**Pull request closed without merging**